[Feature] Feature Overview for Exam Mode

anditurdiu opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
We want to create a page which displays the features of the exam mode in a clear way.
We also made some changes to the exam end page to reduce the clutter. 

### Description
- Create the general structure of the view. 
- Create different routes for students and instructors. 
- Lazy load the features overview module
- Create a pre-configured structure for texts including 1 text, 1 image, 2 texts, 2 images, centre text, first image left, first image right
- Add the features displayed in: 

> https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+German
> https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+English

- Add the text for each feature
- Add the screenshots where relevant

### Coverage
feature-overview.component.ts 67.5%


### Steps for Testing
1. nagivate to /#/features/instructors
2. navigate to /#/features/students
When you hover over a feature, more information is displayed.
When you press on a feature you should automatically scroll to the feature details.

### Screenshots
![image](https://user-images.githubusercontent.com/15110960/95776010-c64ca580-0cc3-11eb-868b-b50f8116089e.png)
![image](https://user-images.githubusercontent.com/15110960/95776027-ce0c4a00-0cc3-11eb-996c-ecd7ba6292e6.png)
![image](https://user-images.githubusercontent.com/15110960/95776041-d49ac180-0cc3-11eb-9898-f0b53bc42bdd.png)

kloessst reviewed (changes requested):
I like the layout and the animation looks cool. Just found a small issue when the short text is displayed again. The line wrapping changes because the text gets bigger. This creates some visual noise.
![wrapIssue](https://user-images.githubusercontent.com/16407766/95852001-5423a180-0d53-11eb-8cda-ec06fabfe211.gif)

In the student view:
- Image for intuitive user interface, user your own IDE are a bit blurry blurry
- LRZ and TUM logo is also a bit blurry

Generally I think some images are too small to really see what's going on e.g. Integrated Anonymous Assessment on instructors view. Maybe a fixed image size would help, or make the images expandable by clicking on them.

Can you only access those pages using the links? Would be nice to have this linked on the landing page in the future.

kokowaahmax reviewed (changes requested):
Well work and really nice done! Only some small suggestions that I think you can make some changes:

------
![Screenshot 2020-10-13 at 11 49 44](https://user-images.githubusercontent.com/48210790/95853210-57b82800-0d55-11eb-96bc-3e3883609efc.png) ![Screenshot 2020-10-13 at 11 49 34](https://user-images.githubusercontent.com/48210790/95853217-58e95500-0d55-11eb-8523-3540800183ae.png) ![Screenshot 2020-10-13 at 12 55 22](https://user-images.githubusercontent.com/48210790/95853218-5981eb80-0d55-11eb-9274-d2a033b3789f.png)

This three cube have different format than other cubes, normally it starts with a verb to discribe an action like "Add a variety of exercises..." But only in these three are in different format

------
![Screenshot 2020-10-13 at 11 54 03](https://user-images.githubusercontent.com/48210790/95853428-b382b100-0d55-11eb-96ce-93ece8d4b023.png)

There are empty space between dividing line, do you want to add an icon inside? If not you can just delete the white space

------

![Screenshot 2020-10-13 at 12 01 58](https://user-images.githubusercontent.com/48210790/95853590-f5135c00-0d55-11eb-9fdc-0a768603f88a.png)

It says the exam mode support file-up load exercise, but the screenshot actually doesn't support

------

![Screenshot 2020-10-13 at 12 06 01](https://user-images.githubusercontent.com/48210790/95853806-491e4080-0d56-11eb-8920-c0dd1727d690.png)

Probably change the alignment from center to leading here would be better, and an entire line with only one word should be avoid

------
![Screenshot 2020-10-13 at 12 03 53](https://user-images.githubusercontent.com/48210790/95854052-ac0fd780-0d56-11eb-960e-9e2d22c4fc28.png)
between sth or sth / between sth and sth?
doing sth or can do sth sounds weird.

![Screenshot 2020-10-13 at 12 13 08](https://user-images.githubusercontent.com/48210790/95854055-aca86e00-0d56-11eb-9bef-0f84daa8cc7a.png)
generated?

MaisiKoleni reviewed (changes requested):
Have some changes for text style and spelling, discussing this directly over Zoom.

anditurdiu commented:
Thank you, everyone, for the review. I added the new version of the texts after discussing it with @MaisiKoleni.
The texts should be free of typos now and better worded than before. If you want to suggest more changes, it is easier to change it here: https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+Page 
using the `Page History` function I can directly see what improvement suggestions you have. This is easier for all of us instead of trying to read through the JSON :D

anditurdiu commented:
> Can you only access those pages using the links? Would be nice to have this linked on the landing page in the future.
**Visual noise:** 
I do not really know a way around the visual noise when you hover over it. What is happening is that the view is shrunk down to 50%, which is why the text also changes. But I think it is fine for now.

**With regards to the image resolutions:**
I will have access tomorrow to a 4k screen so I can take higher resolution screenshots. I'm hoping that will solve the issue. I have set a fixed maximum width of 800px per image. But for some reason it is forcing lower resolution images to scale up which is causing the noise. 

**Access to the feature overview:**
For now it is intended that you can only access it through the link. Maybe in the future we can add it to the initial screen :)

krusche commented:
The exam mode does not support file upload exercises. Please take this into account in the corresponding texts

anditurdiu commented:
> Well work and really nice done! Only some small suggestions that I think you can make some changes:
> 
> ![Screenshot 2020-10-13 at 11 49 44](https://user-images.githubusercontent.com/48210790/95853210-57b82800-0d55-11eb-96bc-3e3883609efc.png) ![Screenshot 2020-10-13 at 11 49 34](https://user-images.githubusercontent.com/48210790/95853217-58e95500-0d55-11eb-8523-3540800183ae.png) ![Screenshot 2020-10-13 at 12 55 22](https://user-images.githubusercontent.com/48210790/95853218-5981eb80-0d55-11eb-9274-d2a033b3789f.png)
> 
> This three cube have different format than other cubes, normally it starts with a verb to discribe an action like "Add a variety of exercises..." But only in these three are in different format
> 
> ![Screenshot 2020-10-13 at 11 54 03](https://user-images.githubusercontent.com/48210790/95853428-b382b100-0d55-11eb-96ce-93ece8d4b023.png)
> 
> There are empty space between dividing line, do you want to add an icon inside? If not you can just delete the white space
> 
> ![Screenshot 2020-10-13 at 12 01 58](https://user-images.githubusercontent.com/48210790/95853590-f5135c00-0d55-11eb-9fdc-0a768603f88a.png)
> 
> It says the exam mode support file-up load exercise, but the screenshot actually doesn't support
> 
> ![Screenshot 2020-10-13 at 12 06 01](https://user-images.githubusercontent.com/48210790/95853806-491e4080-0d56-11eb-8920-c0dd1727d690.png)
> 
> Probably change the alignment from center to leading here would be better, and an entire line with only one word should be avoid
> 
> ![Screenshot 2020-10-13 at 12 03 53](https://user-images.githubusercontent.com/48210790/95854052-ac0fd780-0d56-11eb-960e-9e2d22c4fc28.png)
> between sth or sth / between sth and sth?
> doing sth or can do sth sounds weird.
> 
> ![Screenshot 2020-10-13 at 12 13 08](https://user-images.githubusercontent.com/48210790/95854055-aca86e00-0d56-11eb-9bef-0f84daa8cc7a.png)
> generated?

**title not starting with a verb:**
If you have any suggestion for a different wording I'm happy to change it :) 

**Divider whitespace:**
That was a styling choice on my side, the undivided divider looked strange to me. 

**Alignment**
I can change the center alignment, but I'm not sure you can set a minimum amount of words per line. If you change the width of the window you will always be able to reach a point where a word is pushed to the next line.

anditurdiu commented:
> The exam mode does not support file upload exercises. Please take this into account in the corresponding texts

I removed the file upload mentions

anditurdiu commented:
I added 4k resolution images to replace the old screenshots. I also annotated some of them and used the .gif format for screens which were too long. The images are still displayed as slightly blurry but I think this is due to the resizing effect of .png's which contain fine content such as text.
Is it good enough as it is? Otherwise, I could implement a resizing feature as a follow up where the items expand to full width once a user clicks on them. They could appear as too large though due to the new resolution.

tiffca reviewed (approved):
All looks good, tested on test server 1 on chrome, firefox, safari. 
Read through the confluence page (English) for the wording, nothing major noticed, nice job! 

minor note: I also noticed the discrepancy between "All known Artemis exercise types are supported within the exam mode" and the reality that file upload exercises are not supported. 

Overall it looks really good and functions well!

kloessst commented:
> 
> 
> I added 4k resolution images to replace the old screenshots and annotated. The images are still displayed as slightly blurry but I think this is due to the resizing effect of .png's which contain fine content such as text.
> Is it good enough as it is? Otherwise, I could implement a resizing feature as a follow up where the items expand to full width once a user clicks on them. They could appear as too large though due to the new resolution.
> 
> -Content is also up to date with the latest version on confluence for german and english-

I think this is good for the first iteration. Making the the pictures expandable would be nice in the future. I also think showing some short gifs or videos of the UI would be nice.

krusche commented on src/main/webapp/app/feature-overview/feature-overview.component.ts:
```diff
@@ -0,0 +1,370 @@
+            'featureOverview.students.feature.localIDE.descriptionTextOne',
+            'fa fa-code-fork',
+            undefined,
+            'https://github.com/ls1intum/Artemis/blob/feature/informative-marketing/src/main/resources/public/images/feature-overview/students/clone_repository.png?raw=true',
```
please load those images from the Artemis server and NOT from the Github URL!

kloessst reviewed (approved):
re-approve

jpbernius commented on src/main/webapp/app/feature-overview/feature-overview.component.html:
```diff
@@ -0,0 +1,124 @@
+<head>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
```
Wait, I think it is better if we host this ourselves! Don't we have FA included already?

krusche replied:
yes, please use the local version.
It's really `bad practice` to link to external URLs like this

anditurdiu replied:
see 6d76e007b13b0e7ef7f2599e75802b607d7d5078 & 00b23dde1932cd3def02faff1e6ac637e80c86bc

jpbernius commented on src/main/webapp/app/feature-overview/feature-overview.scss:
```diff
@@ -0,0 +1,264 @@
+@import url('https://fonts.googleapis.com/css2?family=Ubuntu&display=swap');
```
Do we need to load external fonts? What about using the same font setup as elsewhere?

krusche replied:
I agree, please use the same font as in all other places.
It's really `bad practice` to link to external URLs like this

anditurdiu replied:
see 6d76e00

jpbernius reviewed (commented):
Some Images have the test server badge visible or include Grammarly!

FrankeLukas commented:
A few style questions:
Imo it would look better if we could center icon and text, if that is possible.
![grafik](https://user-images.githubusercontent.com/72132281/96151214-76601f80-0f0b-11eb-8870-a514339b866e.png)
would become something like
![grafik](https://user-images.githubusercontent.com/72132281/96151241-7e1fc400-0f0b-11eb-933f-e9e28f931966.png)

We should probably think about reducing the text in some of those previews, since it can look very cramped.

Can we use some sort of block-text format? The following is imo a bit hard to read and look at with all the different line lengths:
![grafik](https://user-images.githubusercontent.com/72132281/96151493-bb845180-0f0b-11eb-84d8-e86a01481822.png)

Some sort of block-text format with shorter line lengths would also help to better fill the space next to images, since the following has a lot of that:
![grafik](https://user-images.githubusercontent.com/72132281/96151663-f7b7b200-0f0b-11eb-9a4e-17b5512390bf.png)

anditurdiu commented:
> A few style questions:
> Imo it would look better if we could center icon and text, if that is possible.
> ![grafik](https://user-images.githubusercontent.com/72132281/96151214-76601f80-0f0b-11eb-8870-a514339b866e.png)
> would become something like
> ![grafik](https://user-images.githubusercontent.com/72132281/96151241-7e1fc400-0f0b-11eb-933f-e9e28f931966.png)
> 
> We should probably think about reducing the text in some of those previews, since it can look very cramped.
> 
> Can we use some sort of block-text format? The following is imo a bit hard to read and look at with all the different line lengths:
> ![grafik](https://user-images.githubusercontent.com/72132281/96151493-bb845180-0f0b-11eb-84d8-e86a01481822.png)
> 
> Some sort of block-text format with shorter line lengths would also help to better fill the space next to images, since the following has a lot of that:
> ![grafik](https://user-images.githubusercontent.com/72132281/96151663-f7b7b200-0f0b-11eb-9a4e-17b5512390bf.png)

I wont be able to implement these suggestions in this PR. I like the block text format, maybe that can be done in a follow-up

anditurdiu commented:
@jpbernius I removed the custom font & the imported icons. Now I this view is using the standard font and the bundled icons. 
@FrankeLukas I implemented the requested changes on the text. I also updated them on the confluence pages:

- https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+German
- https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+English

anditurdiu commented on src/main/webapp/i18n/en/featureOverview.json:
```diff
@@ -0,0 +1,139 @@
+                "createConductAssess": {
+                    "title": "Create, Conduct and Assess Exams",
+                    "shortDescription": "Create, conduct and assess, all within the Artemis platform.",
+                    "descriptionTextOne": "Artemis now supports the exam mode, which offers a unique way of assessing the performance of your students. Yea can create an unlimited amount of exams for each course and you can register all or some of your students. Within the exam, you can include an unlimited amount of exercises like from the exercise types of the Artemis platform including program ...
```
see 6d76e00

anditurdiu commented on src/main/webapp/i18n/en/featureOverview.json:
```diff
@@ -0,0 +1,139 @@
+                "configuration": {
+                    "title": "Fully Configurable",
+                    "shortDescription": "Bulk register students from .csv. Configure the exam to your needs. Edit working time of individual students.",
+                    "descriptionTextOne": "Artemis allows you to configure all aspects of the exam. You can change the name of the exam and set the module number, examiner and course name. Furthermore, You can preset the timeline of the exam to fit your needs by deciding when the exam will become visible to the students, when the working time begins and ends and when the results are published. Ad ...
```
See 6d76e00

anditurdiu commented on src/main/webapp/i18n/en/featureOverview.json:
```diff
@@ -0,0 +1,139 @@
+                "exerciseTypes": {
+                    "title": "Multiple Exercise Types",
+                    "shortDescription": "Add a variety of exercises, like programming, modeling, text and multiple-choice questions.",
+                    "descriptionTextOne": "Most Artemis exercise types are supported within the exam mode. These include programming-, modeling-, text- and quiz exercises such as multiple choice, drag and drop and short answer questions. You can an unlimited number of exercises to an exam.\n\nFor programming exercises, students decide between working with the integrated code editor or can clone t ...
```
See 6d76e00

MaisiKoleni reviewed (changes requested):
Looks very good, here are the last few errors I found:

- Students (EN):
   1. Exam Summary, at the end 
       > Artemis offers **to** option [...]
- Students (DE)
   1. Offline Weiterarbeiten
      > Für Programmieraufgaben **,** brauchst du nur [...]
   2. Qualität und Fairness der Bewertung
       > Diese automatische Bewertung kann durch **einer** manuellen Bewertung [...]
       
       and
       > Diese Funktion kann für Text- und **Modellierübungen** [...]
       
       we used "Modellierungsaufgaben" everywhere else.
       In general, there are a few places where we use "übungen", we should replace all with "aufgaben"
- Instructors (EN)
   1. Conduct Test Runs, at the end
       > Test runs are also a great way to **familiarise** [...]
   2. Automatic Assessment, at the end
       > It then suggests how elements of the submissions should be evaluated **during the course of further correction**.
       
       This is very strange/wordy, it would shorten that to "during further correction" or similar.
   3. Review Individual Exams
       > [...] as an instructor you have **to** ability to access [...]
       
       and
       > You can use this feature to review **indivdual** student exams [...]
- Instructors (DE)
   1. Klausur Erstellen, Durchführen und Bewerten
       > [...] einschließlich Programmier-, Modellierungs-, Text- und Quiz-**Übungen** [...]
   2. Automatische Korrektur, near the end
       > Hier lernt Artemis mittels Maschinelles Lernen [...]
       > Hier lernt Artemis mittels maschinellem Lernens [...]

anditurdiu commented:
> 1. This is the only occurrence I found here.

See 21aea42692493096128337cd3f45e0ce61905236

MaisiKoleni reviewed (approved):
Looks good to me now. There are more points that could be improved (the German is still a bit strange at some places), but I think we managed to remove almost all errors now.
